crates/typst-library/src/compute/construct.rs

```diff
@@ -528,6 +531,71 @@ cast! {
     },
 }
 
+/// Creates a version.
+///
+/// Arguments are the version's components;
+/// Array arguments are flattened.
+///
+/// ## Example { #example }
+/// ```example
+/// // any number of components is ok
+/// #let version0 = version()
+/// #let version1 = version(1)
+/// #let version5 = version(1, 2, 3, 4, 5)
+///
+/// #let version5_from_array = version((1, 2, 3, 4, 5))
+/// #let version5_mixed = version(1, (2, 3), 4, (5,))
+/// ```
+///
+/// Display: version
+/// Category: construct
+#[func]
+pub fn version(
+    /// The components of the version (array arguments are flattened)
+    #[variadic]
+    components: Vec<Spanned<VersionComponents>>,
+) -> SourceResult<Version> {
+    let mut res = Version::new();
+    let mut errs = vec![];
+
+    let mut push = |i, span| match VersionComponent::new(i) {
+        Some(i) => res.push(i),
+        None => errs.push(SourceDiagnostic::error(
+            span,
+            "version component must be non-negative",
```

Comment:
You could enforce this on the type system level by having `VersionComponents` be `u64` or something else.

Reply:
Well I tried that (that's what `VersionComponent` (singular!) is, after all), but I failed to make it implement `Reflect`. Any ideas?

Reply:
You don't need to implement `Reflect` yourself, the `cast!` macro will do it for you. Just replace `i: i64` in `cast!` with `i: u32` or `i: u64`.

Reply:
Ohhhh, alright, let me try it.